### PR TITLE
Symfony 5 Controller dependency injection

### PR DESCRIPTION
**Pull request by davidmpaz**
Hi,

thanks for this project. I would like to use it in a Symfony 5 project i have but i am finding some issues regarding package dependencies and also DI Container.

After a small test i wonder if i can introduce some more configuration inside the bundle. To understand what i mean, i had to do this bindings in my application services.yaml file:

```yaml
    Payum\Bundle\PayumBundle\Controller\:
        resource: '../vendor/payum/payum-bundle/Controller/'
        tags: ['controller.service_arguments']
    Payum\Core\Registry\RegistryInterface:
        public: true
        alias: 'payum'
```
Can i put these also inside the bundle to reduce effort on integration later on in applications using it?

Could someone please check this PR?  Provide feedback or merge it?  /cc @Tetragramat, @bennukem, @joshbmarshall ?

I am trying to solve issue: https://github.com/Payum/PayumBundle/issues/507

Thanks in advance,
David

### CODE REVIEW

**Comment by Tetragramat** [comment]
There are already 2 other PRs for this issue. See https://github.com/Payum/PayumBundle/pull/513 and https://github.com/Payum/PayumBundle/pull/509.

Before fixing this issue, those controller tests should be modified so they test that controller services are properly configured.
Right now it's skipped. I'll look into this in next days.

> Can i put these also inside the bundle to reduce effort on integration later on in applications using it?

Yes all services should be configured out of the box including controllers.

**Comment by davidmpaz** [comment]
Hi there,

thank for your comments.

>There are already 2 other PRs for this issue. See #513 and #509.

I noticed now about those others. I am trying to find the common denominator on the 3 to find a solution that solves all approaches. 

As i see @BoShurik says his PR is an alternative solution to your PR. In both of them the container is injected via setter method. I assume this is only done in order to get the `payum` instance from container later on. Am i correct here? 

Is the above is the only reason for injecting the container, i think the solution proposed in this PR will serve better to [Symfony best practices](https://symfony.com/doc/current/best_practices.html#use-dependency-injection-to-get-services) since it inject the Payum instance and avoid depending on the container to get it.

The only work missing here would be to move the configuration i already have in the application into the Bundle configuration.

What do you think?

>Before fixing this issue, those controller tests should be modified so they test that controller services are properly configured. Right now it's skipped.

Could you please point me which are those, so i can take a look to them.

thanks in advance for checking this out.
Best regards,
David

**Comment by Tetragramat** [comment]
> Could you please point me which are those, so i can take a look to them.

Functional tests of controller actions https://symfony.com/doc/4.4/testing.html#functional-tests
If you have no idea how to make them then I'll do it arround tuesday.

**Comment by davidmpaz** [comment]
Hey, thanks for your message.

What i understood from here:
>Before fixing this issue, those controller tests should be modified...

was that there exist some tests already. I wanted to know which ones needs to be modified to look at them.

I do know those Functional Tests in the documentation you sent. What i don't know is what to test already :). I think it is a good idea to wait for your additions as you suggested.

Best regards,
David

**Comment by Tetragramat** [comment]
I'm sorry If my responses feel a bit chaotic. I was talking about controller tests you've modified. But after revieving other PR for this issue I've noticed that functional tests are separated in it's own directory. So instead of you should make functional tests in  `Tests/Functional/Controller/` dir. Same as in my PR https://github.com/Payum/PayumBundle/pull/509/files. You can copy and modify those.

**Comment by davidmpaz** [comment]
Hey 👋, no problem at all. Actually i should apologise with you since last week i had some time to invest on the project and i feel like i was spamming you with mentions on my comments. We are all busy and with little time, so it is fine.

I think i got a better insight now about the DI in the bundle and controllers. I did notice that in either case we need to configure the container in the controllers if we want to make some real functional tests, like WebTestCase. 

Some how i played with (and i liked) the idea of some kind of "late binding". Like a flex recipe or alike that could add the configuration needed at install time to configure controllers and bind `payum` to `RegistryInterface` like i did for my application.

In any case, i did merge your PR and add some changes to allow the injection on constructor. I also did leave the Tests under: `Tests/Controller` in their place since they are plain Unit Tests, all is mocked there, they are not functional test really.

Best,
David

File: Resources/config/controller.xml [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="Payum\Bundle\PayumBundle\Controller\AuthorizeController" class="Payum\Bundle\PayumBundle\Controller\AuthorizeController" public="true">
+            <call method="setContainer">
+                <argument type="service" id="service_container" />
+            </call>
+            <argument type="service" id="payum"/>
```
Comment (Tetragramat) [inline Resources/config/controller.xml:12]: Each controller service should have `<tag name="container.service_subscriber" />`. So only reduced container is injected.

File: Resources/config/payum.xml [hunk 2]
```diff
@@ -49,7 +49,9 @@
         <service id="payum" class="Payum\Core\Payum">
             <factory service="payum.builder" method="getPayum" />
         </service>
-        <service id="Payum\Core\Payum" alias="payum"></service>
+        <service id="Payum\Core\Payum" alias="payum"/>
+        <!-- Allows to bind Payum as RegistryInterface also -->
+        <service id="Payum\Core\Registry\RegistryInterface" alias="Payum\Core\Payum"/>
```
Comment (Tetragramat) [inline Resources/config/payum.xml:54]: I don't think this is good idea. I'm not totally sure, but it may cause issues. At least it should be in it's own PR.

File: Controller/PayumController.php [hunk 3]
```diff
@@ -2,18 +2,33 @@
...
+    /**
+     * @var Payum
+     */
+    private $payum;
+
+    /**
+     * PayumController constructor.
+     * @param RegistryInterface $payum
+     */
+    public function __construct(RegistryInterface $payum)
```
Comment (Tetragramat) [inline Controller/PayumController.php:21]: There needs to be `Payum\Core\Payum` typehint. `Payum\Core\Registry\RegistryInterface` does not contain all methods controllers call.

File: Resources/config/payum.xml (hunk 2 above)
Comment (davidmpaz) [reply Resources/config/payum.xml:54]: If we follow the idea of below comment, then this is not needed. I will remove it

File: Controller/PayumController.php (hunk 3 above)
Comment (Tetragramat) [reply Controller/PayumController.php:21]: Still waiting for this.

**Review by Tetragramat — CHANGES_REQUESTED** [review]

File: Controller/PayumController.php (hunk 3 above)
Comment (davidmpaz) [reply Controller/PayumController.php]: Oops that was clumsy ;) ... done

**Review by Tetragramat — APPROVED** [review]
@makasim This PR is ok.

**Comment by davidmpaz** [comment]
Hi @makasim, could we please merge this one?

thanks in advance
regards,
David

**Comment by dpfaffenbauer** [comment]
@makasim Can you please merge that? Thanks.

File: Tests/Controller/AbstractControllerTest.php [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    protected const AFTER_URL = 'http://example.com/theAfterUrl';
+
+    protected $token;
+    protected $httpRequestVerifierMock;
+    protected $gatewayMock;
+    protected $registryMock;
+    protected $payum;
+    protected $request;
+
+    protected function initMocks()
```
Comment (pierredup) [inline Tests/Controller/AbstractControllerTest.php:26]: This can be changed to `setUp` so that it is automatically executed for each test and so that you don't have to call this method each time in the tests

File: Resources/config/controller.xml [hunk 5]
```diff
@@ -0,0 +1,64 @@
+<?xml version="1.0" encoding="UTF-8" ?>
+
+<container xmlns="http://symfony.com/schema/dic/services"
+           xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
+           xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
+
+    <services>
+        <service id="Payum\Bundle\PayumBundle\Controller\AuthorizeController" class="Payum\Bundle\PayumBundle\Controller\AuthorizeController" public="true">
```
Comment (pierredup) [inline Resources/config/controller.xml:8]: This config can be simplified to automatically register all controllers without having to manually specify all of them

```xml
<prototype namespace="Payum\Bundle\PayumBundle\Controller\" resource="../../Controller/*" public="true">
    <tag name="controller.service_arguments"/>
</prototype>
```

You can then also enable auto-wiring so that the right services are injected by default

```xml
<defaults autowire="true" />
```

File: Tests/Functional/Controller/AuthorizeControllerTest.php [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+namespace Payum\Bundle\PayumBundle\Tests\Functional\Controller;
+
+use Payum\Bundle\PayumBundle\Tests\Functional\WebTestCase;
+use Symfony\Component\HttpKernel\Exception\NotFoundHttpException;
+
+class AuthorizeControllerTest extends WebTestCase
+{
+    /**
+     * @ticket 507
```
Comment (pierredup) [inline Tests/Functional/Controller/AuthorizeControllerTest.php:11]: I don't think this comment provides anything useful.

File: Tests/Controller/CaptureControllerTest.php [hunk 7]
```diff
@@ -38,7 +38,18 @@ public function shouldBeSubClassOfController()
      */
     public function throwBadRequestIfSessionNotStartedOnDoSessionAction()
     {
-        $controller = new CaptureController;
+        $this->registryMock = $this->createMock(RegistryInterface::class);
+        $this->httpRequestVerifierMock = $this->createMock(
+            HttpRequestVerifierInterface::class
+        );
+        $this->payum = new Payum(
```
Comment (pierredup) [inline Tests/Controller/CaptureControllerTest.php:45]: This should already be handled in the `AbstractControllerTest`, so it's not necessary to re-define the mock again here (Once the `initMocks` is updated to `setUp`)

**Review by pierredup — CHANGES_REQUESTED** [review]

File: Resources/config/controller.xml (hunk 5 above)
Comment (Tetragramat) [reply Resources/config/controller.xml:8]: You can put that in project but you should not do that in reusable bundles. https://symfony.com/doc/current/bundles/best_practices.html#services

File: Tests/Functional/Controller/AuthorizeControllerTest.php (hunk 6 above)
Comment (Tetragramat) [reply Tests/Functional/Controller/AuthorizeControllerTest.php:11]: allows to filter tests based on their issue ID in phpunit which is very usefull while writing test and later it provides link to issue where you can often find explanation to why is this test here and what it is supposed to do.

Comment (pierredup) [reply Tests/Functional/Controller/AuthorizeControllerTest.php:11]: Sure, but I think it's only useful in tests covering specific edge cases. Also, only having the ticket id means you have to find the repo and issue number manually, and it might not contain enough information to explain the purpose of the test.

So this annotation can be accompanied by a `@see https://github.com/Payum/PayumBundle/issues/507` to link to the issue directly

**Review by pierredup — APPROVED** [review]

**Comment by pierredup** [comment]
Thank you @davidmpaz
